File: summary/go/16_eratosthenes.go
package main

import "fmt"

func sieveOfEratosthenes(max int) []int {
	numbers := make([]bool, max+1)
	for i := 2; i <= max; i++ {
		numbers[i] = true
	}

	p := 2
	for p*p <= max {
		if numbers[p] {
			for i := p * p; i <= max; i += p {
				numbers[i] = false
			}
		}
		p++
	}

	var primes []int
	for i := 2; i <= max; i++ {
		if numbers[i] {
			primes = append(primes, i)
		}
	}

	return primes
}

func main() {
	maxNumber := 30
	primes := sieveOfEratosthenes(maxNumber)
	fmt.Println("Primes up to", maxNumber, ":", primes)
}

File: summary/c++/16_eratosthenes.cpp
#include <iostream>
#include <vector>

std::vector<int> sieve_of_eratosthenes(int max) {
    std::vector<bool> is_prime(max + 1, true);
    std::vector<int> primes;
    is_prime[0] = is_prime[1] = false; // 0 and 1 are not prime numbers

    for (int num = 2; num * num <= max; ++num) {
        if (is_prime[num]) {
            for (int multiple = num * num; multiple <= max; multiple += num) {
                is_prime[multiple] = false;
            }
        }
    }

    for (int num = 2; num <= max; ++num) {
        if (is_prime[num]) {
            primes.push_back(num);
        }
    }

    return primes;
}

int main() {
    int max_number = 30;
    std::vector<int> primes = sieve_of_eratosthenes(max_number);
    std::cout << "Primes up to " << max_number << ": ";
    for (int prime : primes) {
        std::cout << prime << " ";
    }
    std::cout << std::endl;
    return 0;
}
